### PR TITLE
perf(compare): stream keyed compare diff instead of full keyed maps

### LINKED ISSUE #704: [PERF] follow up on #600: keep --compare --compare-key usable on large tables
## Summary

Issue #600 improved the compare path so both full tables do not need to stay live at the same time, but keyed compare is still expensive on large or wide inputs.

## Problem

`--compare --compare-key` still builds large keyed row maps in Go after loading table rows. That keeps the command convenient for small datasets, but it can become memory-heavy for wide tables or large CSV revisions.

## Why this matters

`--compare` is an existing usability feature for diffing datasets quickly. If it falls over on realistic input sizes, users have to abandon the workflow and drop down to custom SQL or external tools.

## Proposal

Continue the performance work from #600 so keyed compare remains practical on larger datasets. Possible directions include:
- using SQL to pre-filter or join keyed rows before building the final report
- reducing the amount of row data copied into Go maps
- streaming or chunking the keyed diff when feasible

The goal is not to change the compare feature set, only to keep the current feature usable at larger scales.

## Acceptance criteria

- large keyed compares avoid unnecessary full-row duplication in memory
- current JSON and text report semantics stay stable
- CLI input order and current compare behavior remain unchanged
- add a benchmark or regression test that exercises a larger keyed compare workload

### PR DESCRIPTION
**Pull request by nao1215**
Follow-up on #600. Keyed compare (`--compare --compare-key`) still loaded both full tables into Go keyed maps after materializing the whole `SELECT *` result, copying every row into a `map[string]*string`. For wide or large CSV revisions that holds two full copies in memory even when only a handful of rows actually differ.

This change pushes the keyed diff into the in-memory SQLite database through the existing streaming query path. A first pass over the left table records only a per-key row fingerprint, not the row data. The pass over the right table compares each row against that fingerprint on the fly: a key absent on the left is added, a fingerprint mismatch is a modification candidate, and only those right rows are retained. A final left pass fetches just the rows behind the removed and candidate keys. Unchanged rows never enter Go, so peak memory scales with the size of the diff and the key set rather than with both full tables.

Correctness is preserved exactly. Fingerprints are only a fast inequality filter; every candidate is re-checked with the same in-memory `rowMapsEqual`, so a fingerprint collision can never report a false change. The key string still maps a SQL NULL key onto the empty string (so a NULL key and an empty-string key collide as before), duplicate keys are still rejected in row order with the same message, the case-insensitive key resolution is kept, and the differing-schema rule (every shared row modified) is reproduced. The JSON and text report output is byte-for-byte unchanged, and CLI input order and compare behavior are unchanged.

Tests added: a regression test that pins the SQL path to the in-memory oracle on a fixed dataset, a 200-case metamorphic property test that asserts the SQL path matches the oracle on random keyed datasets (digits, letters, empty strings, Unicode), a wide-table benchmark (`BenchmarkBuildCompareReportKeyedWide`) with `b.ReportAllocs()` for the large keyed workload, and shellspec cases for a duplicate-key rejection and a larger keyed diff through the built binary.

Benchmark before/after (`buildCompareReport`, keyed, `-benchtime 50x`):

Narrow 5000x3: 7.07 MB/op -> 4.03 MB/op (-43%).
Wide 5000x21: 27.75 MB/op -> 19.36 MB/op (-30%).

Total bytes allocated drop in both cases; more importantly the rows retained for the report fall from the whole table to just the diff, which is the peak-memory win the issue targets.

Closes #704

## Summary by CodeRabbit

* **Performance**
  * Enhanced `--compare --compare-key` to compute keyed diffs in SQLite using a streaming approach, materializing only added/removed/modified rows to reduce memory usage.
* **Bug Fixes**
  * Improved keyed diff correctness, including proper NULL handling vs empty strings and more reliable detection of modified rows.
  * Duplicate keys for the compare key now fail with a clear “not unique” error.
* **Tests**
  * Added regression tests covering delimiter-collision scenarios, SQLite-vs-in-memory parity, NULL/ordering edge cases, and wide-table benchmarks.

### CODE REVIEW

**Comment by nao1215** [comment]
@coderabbitai review

File: shell/compare.go [hunk 1]
```diff
@@ -242,42 +243,246 @@ func (s *Shell) buildCompareReport(ctx context.Context, left, right, key string)
...
+		if _, dup := seen[k]; dup {
+			return duplicateKeyError(rightKeyCol, right, k)
+		}
+		seen[k] = struct{}{}
+		lfp, shared := leftFP[k]
+		switch {
+		case !shared:
+			addedRows[k] = recordToCompareRow(rightCols, record, nulls)
+		case !columnsMatch || lfp != rowFingerprint(rightCols, record, nulls):
+			candidateRows[k] = recordToCompareRow(rightCols, record, nulls)
```
[the remark this answers is not in the document]
Comment (nao1215) [reply shell/compare.go]: Fixed in 0092261. Replaced the fingerprint-based modified detection with an exact SQL comparison. Modified keys are now found by joining the two tables on the key expression and keeping rows where any shared column is not NULL-safe equal on its TEXT projection (CAST(col AS TEXT) IS NOT CAST(col AS TEXT)), which mirrors rowMapsEqual exactly. The delimiter/hash collision can no longer hide a real change. A temporary expression index on each side keeps the join roughly O(rows) (dropped afterward, so no schema changes), and rows are still fetched only for added/removed/modified keys, so the diff stays memory-bounded. Added an adversarial regression test with the exact ["a","b\x00\x01c"] vs ["a\x00\x01b","c"] case asserting it is reported modified and matches the oracle.

File: shell/compare.go [hunk 2]
```diff
@@ -242,42 +243,246 @@ func (s *Shell) buildCompareReport(ctx context.Context, left, right, key string)
...
+func (s *Shell) streamKeyedRows(ctx context.Context, name, keyCol string, cols []inspectColumn, fn func(key string, record []string, nulls []bool) error) error {
+	quote := s.usecases.importer.QuoteIdentifier
+	quotedTable := quote(name)
+	quotedKey := quote(keyCol)
+	keyIdx := len(cols) // the appended key expression is the last column
+	return s.usecases.query.QueryStream(ctx,
+		"SELECT *, "+keyExpr(quotedKey)+" FROM "+quotedTable,
+		func(record []string, nulls []bool) error {
+			if keyIdx >= len(record) {
+				return nil
```
[the remark this answers is not in the document]
Comment (nao1215) [reply shell/compare.go]: Fixed in 0092261. streamKeyedRows now returns fmt.Errorf("compare key expression missing while streaming table %s", name) when the appended key column is missing, instead of returning nil and silently dropping the row, so a malformed stream fails loudly.

File: shell/compare.go [hunk 3]
```diff
@@ -242,42 +242,319 @@ func (s *Shell) buildCompareReport(ctx context.Context, left, right, key string)
...
+	}
+	if _, err := s.usecases.query.Exec(ctx,
+		"CREATE INDEX IF NOT EXISTS "+quote(rightIdx)+" ON "+quote(right)+" ("+keyExpr(quote(rightKeyCol))+")"); err != nil {
+		_, _ = s.usecases.query.Exec(ctx, "DROP INDEX IF EXISTS "+quote(leftIdx))
+		return func() {}, fmt.Errorf("failed to index compare key of %s: %w", right, err)
+	}
+	return func() {
+		_, _ = s.usecases.query.Exec(ctx, "DROP INDEX IF EXISTS "+quote(leftIdx))
+		_, _ = s.usecases.query.Exec(ctx, "DROP INDEX IF EXISTS "+quote(rightIdx))
+	}, nil
```
[the remark this answers is not in the document]
Comment (nao1215) [reply shell/compare.go]: Fixed in b293588. The index name is now unique per call (sanitized table name plus a process-wide atomic counter), so it cannot collide with a user index. CREATE no longer uses IF NOT EXISTS, so an unexpected name clash fails loudly instead of silently reusing another index. The cleanup drops only the indexes this call created, and a drop failure is now warned to config.Stderr rather than discarded. The indexes are still dropped after the compare, so a read-only compare leaves no schema changes. Added TestBuildCompareReport_PreservesUserIndex asserting a user index sharing the base name survives a keyed compare.

File: shell/compare_test.go [hunk 4]
```diff
@@ -390,6 +393,230 @@ func TestBuildCompareReport_KeyedDiff(t *testing.T) {
...
+		var b strings.Builder
+		w := csv.NewWriter(&b)
+		_ = w.Write([]string{"id", "a", "b"})
+		for _, row := range rows {
+			_ = w.Write(row)
+		}
+		w.Flush()
+		if err := os.WriteFile(path, []byte(b.String()), 0o600); err != nil {
+			t.Fatal(err)
+		}
```
[the remark this answers is not in the document]
Comment (nao1215) [reply shell/compare_test.go:462]: Fixed in b293588. Both compare test fixture helpers (writeRaw and writeCSV) now check each csv.Writer.Write return value and call w.Error() after Flush, failing the test with t.Fatal on any error, instead of discarding them.
